core/src/main/java/org/mule/runtime/core/execution/ModuleFlowProcessingPhase.java

```diff
@@ -64,118 +66,182 @@ public boolean supportsTemplate(MessageProcessTemplate messageProcessTemplate) {
   @Override
   public void runPhase(final ModuleFlowProcessingPhaseTemplate template, final MessageProcessContext messageProcessContext,
                        final PhaseResultNotifier phaseResultNotifier) {
-    Work flowExecutionWork = new Work() {
 
-      @Override
-      public void release() {
+    try {
 
-      }
+      final MessagingExceptionHandler exceptionHandler = messageProcessContext.getFlowConstruct().getExceptionListener();
+      MessageSource messageSource = messageProcessContext.getMessageSource();
+      ComponentIdentifier sourceIdentifier = messageProcessContext.getSourceIdentifier();
+      Consumer<MessagingException> errorConsumer =
+          getErrorConsumer(messageSource, template.getFailedExecutionResponseParametersFunction(),
+                           messageProcessContext, template, phaseResultNotifier);
+
+      Event templateEvent = Event
+          .builder(create(messageProcessContext.getFlowConstruct(), sourceIdentifier.getNamespace()))
+          .message((InternalMessage) template.getMessage()).build();
+
+      // TODO MULE-11167 Policies should be non blocking
+      if (System.getProperty("enableSourcePolicies") == null) {
+
+        just(templateEvent)
+            .doOnNext(request -> fireNotification(messageProcessContext.getMessageSource(), request,
+                                                  messageProcessContext.getFlowConstruct(),
+                                                  MESSAGE_RECEIVED))
+            .then(request -> from(template.routeEventAsync(request)))
+            .doOnSuccess(getSuccessConsumer(messageSource, templateEvent, exceptionHandler, errorConsumer,
```

Comment:
I know, this class is already in desperate need to refactor and unit tests. I gave this feedback already in a previous PR. This here is temporary until we sort out non-blocking polcies at which point this whole class needs a clean up and addition of unit tess.